pep-9999.rst

```diff
@@ -0,0 +1,341 @@
+imported from it) is actually referenced during execution. "Transparent" means
+that besides the delayed import (and necessarily observable effects of that,
+such as delayed import side effects and changes to ``sys.modules``), there is
+no other observable change in behavior: the imported object is present in the
+module namespace as normal and is transparently loaded whenever first used: its
+status as a "lazy imported object" is not directly observable.
+
+The requirement that the imported object be present in the module namespace as
+usual, even before the import has actually occurred, means that we need some
+kind of "lazy object" placeholder to represent the not-yet-imported object.
+The transparency requirement dictates that this placeholder must never be
+visible to Python code; any reference to it must trigger the import and replace
+it with the real imported object.
+
+Given the possibility that Python (or C extension) code may pull objects
+directly out of a module ``__dict__``, the only way to reliably prevent
+accidental leakage of lazy objects is to have the dictionary itself be
+responsible to ensure resolution of lazy objects on lookup. 
+
+To avoid a performance penalty on the vast majority of dictionaries which never
+contain any lazy object, we install a specialized lookup function
+(``lookdict_unicode_lazy``) for module namespace dictionaries when they first
+gain a lazy-object value. This lookup function checks the looked-up value and
+resolves it before returning it, if it is a lazy object.
+
+This implementation comprehensively prevents leakage of lazy objects, ensuring
+they are always resolved to the real imported object before anyone can get hold
+of them for any use, while avoiding any noticeable performance impact on
+dictionaries in general.
+
+
+Specification
+=============
+
+Lazy imports are opt-in, and globally enabled via a new ``-L`` flag to the
+python interpreter, or a ``PYTHONLAZYIMPORTS`` environment variable.
+
+When enabled, the loading and execution of all (and only) top level imports is
+deferred until the imported name is used. This could happen immediately (e.g.
+on the very next line after the import statement) or much later (e.g. while
+using the name inside a function being called by some other code at some later
+time.)
+
+For these top level imports, there are two exceptions which will make them
+eager (not lazy): imports inside ``try``/``except``/``finally``  or ``with``
+blocks, and star imports (``from foo import *``.) Imports inside
+exception-handling blocks remain eager so that any exceptions arising from the
+import can be handled. Star imports must remain eager since performing the
+import is the only way to know which names should be added to the namespace.
+
+Imports inside class definitions or inside functions/methods are not "top
+level" and are never lazy.
+
+Dynamic imports using ``__import__`` or ``importlib.import_module`` are also
+never lazy.
```

Comment:
I don't exactly remember the markup, but I think you can reference the functions in the stdlib documentation.

Also, I've talked with some folks about whether `__import__()` or `import_module()` should grow an option to enable lazy import.  The PEP should at least discuss the topic, and promote a particular opinion about that (if it doesn't already - I'm reading this linearly).

Reply:
> I think you can reference the functions in the stdlib documentation.

Got any pointer here? I couldn't find anything about this in PEP 12, nor could I find an example of another PEP doing this. (I was grepping for things like `:func` and `:meth`, I may just not know the right thing to grep for.)

Reply:
You can't at the moment, sorry.

Reply:
> You can't at the moment, sorry.

Ah, okay.  Could you please add parentheses then, to make it clear you're talking about `__import__()` and `importlib.import_module()`?